### PR TITLE
feat(date): add location support to date.truncate function

### LINKED ISSUE #4656: Location support for date.truncate function
At the moment, truncate() does the truncation in UTC, e.g.:

```
option location = timezone.location(name: "Europe/Madrid")

date.truncate(t: 2019-06-03T13:59:01Z, unit: 1d) => 2019-06-03T00:00:00Z
date.truncate(t: 2019-06-03T13:59:01Z, unit: 1mo) => 2019-06-01T00:00:00Z
date.truncate(t: 2019-06-03T13:59:01Z, unit: 1y) => 2019-01-01T00:00:00Z
```

However, I would have expected to take into account the current location. For instance, in "Europe/Madrid" "2019-06-03T00:00:00" and "2019-06-01T00:00:00" both lands in CEST (+02:00) but the beginning (2019-01-01) is at CET (+01:00). Thus I would expect the following location-aware behavior:

```
option location = timezone.location(name: "Europe/Madrid")

date.truncate(t: 2019-06-03T13:59:01Z, unit: 1d) => 2019-06-02T22:00:00Z
date.truncate(t: 2019-06-03T13:59:01Z, unit: 1mo) => 2019-05-31T22:00:00Z
date.truncate(t: 2019-06-03T13:59:01Z, unit: 1y) => 2018-12-31T23:00:00Z
```
To elaborate further, this is useful in scenarios similar to https://chrispnerd.blogspot.com/2021/12/i-have-application-that-feeds-periodic.html but in which you want to find precise location-aware boundaries.

Without a way to do location-aware truncation, location-aware windowing can create "leftover" windows due to the UTC vs local-tzone mismatches.

### PR DESCRIPTION
**Pull request by 2opremio**
1. Added support for location option to date.truncate by replacing the use of `execute.Window` by `interval.Window`
2. Go & acceptance test added in date package

Inspired by #4327 

fixes #4656 

### Done checklist
- [ ] docs/SPEC.md updated
- [x] Test cases written

### CODE REVIEW

**Comment by 2opremio** [comment]
CC @skartikey

File: stdlib/date/truncate_duration_test.flux [hunk 1]
```diff
@@ -33,12 +45,38 @@ outData =
...
+    }
+
+testcase truncate_duration_location {
+        option location = timezone.location(name: "Europe/Madrid")
+
+        got =
+            csv.from(csv: inData)
+                |> range(start: 2018-05-22T19:53:00Z)
+                |> drop(columns: ["_start", "_stop"])
+                |> map(fn: (r) => ({r with _value: int(v: date.truncate(t: duration(v: r._value), unit: 1mo))}))
```
Comment (Marwes) [inline stdlib/date/truncate_duration_test.flux:61]: The existing test also used `int(v: ...)` here which is why I assume you did but would it not be easier to read the expected output if we omitted the integer conversion?

File: stdlib/date/truncate_duration_test.flux [hunk 2]
```diff
@@ -33,12 +45,38 @@ outData =
...
+,,0,FF,_m,2018-05-22T19:53:20.000000000Z,1890774000000000000
+,,0,FF,_m,2018-05-22T19:53:30.000000000Z,1893452400000000000
+,,0,FF,_m,2018-05-22T19:53:40.000000000Z,1893452400000000000
+,,0,FF,_m,2018-05-22T19:53:50.000000000Z,1893452400000000000
+,,0,FF,_m,2018-05-22T19:54:00.000000000Z,1893452400000000000
+",
+            )
+
+        testing.diff(got: got, want: want)
+    }
```
Comment (Marwes) [inline stdlib/date/truncate_duration_test.flux:82]: Nit: Missing newline at the end of file (usually your editor would add one in my experience :shrug: )

**Review by Marwes — CHANGES_REQUESTED** [review]

File: stdlib/date/date.go [hunk 3]
```diff
@@ -261,12 +262,20 @@ func init() {
...
-				return values.NewTime(b.Start), nil
+				intervalLocation, err := interval.LoadLocation(location)
+				if err != nil {
+					return nil, err
+				}
+				w, err := interval.NewWindowInLocation(u.Duration(), u.Duration(), values.Duration{}, intervalLocation)
+				if err != nil {
+					return nil, err
+				}
+				b := w.GetLatestBounds(t)
```
[the remark this answers is not in the document]
Comment (2opremio) [reply stdlib/date/date.go:277]: Agreed! BTW `execute.Window` seems to only be used by `interpolate`. So I think may be able to get rid of it and completely replace it by `interval.Window`.

File: stdlib/date/date.go [hunk 4]
```diff
@@ -240,7 +241,7 @@ func init() {
 		),
 		"truncate": values.NewFunction(
 			"truncate",
-			runtime.MustLookupBuiltinType("date", "truncate"),
+			runtime.MustLookupBuiltinType("date", "_truncate"),
```
[the remark this answers is not in the document]
Comment (2opremio) [reply stdlib/date/date.go:244]: I don't think it's necessary, I followed the pattern from https://github.com/influxdata/flux/pull/4327 for consistency.

Also, wouldn't the tests fail if that was required?

File: stdlib/date/truncate_duration_test.flux (hunk 1 above)
Comment (2opremio) [reply stdlib/date/truncate_duration_test.flux:61]: I tired to keep it consistent but I can just use a timestamp. I will change it.

File: stdlib/date/truncate_duration_test.flux (hunk 2 above)
Comment (2opremio) [reply stdlib/date/truncate_duration_test.flux:82]: I will address this too 👍

Comment (2opremio) [reply stdlib/date/truncate_duration_test.flux:82]: > usually your editor would add one in my experience 

Not for unknown filetypes (at least in Golang :) )

**Comment by 2opremio** [comment]
@Marwes PTAL!

**Comment by 2opremio** [comment]
@Marwes friendly ping

**Review by Marwes — APPROVED** [review]

**Pull request merged**